homeassistant/components/rainmachine/select.py

```
@@ -0,0 +1,165 @@
+                and coordinator.data
+                and key_exists(coordinator.data, description.data_key)
+            )
+        ]
+    )
+
+
+class FreezeProtectionTemperatureSelect(RainMachineEntity, SelectEntity):
+    """Define a RainMachine select."""
+
+    entity_description: FreezeProtectionSelectDescription
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        data: RainMachineData,
+        description: FreezeProtectionSelectDescription,
+        unit_system: str,
+    ) -> None:
+        """Initialize."""
+        super().__init__(entry, data, description)
+
+        self._api_value_to_label_map = {}
+        self._label_to_api_value_map = {}
+
+        for option in description.options:
+            if unit_system == CONF_UNIT_SYSTEM_IMPERIAL:
+                label = option.imperial_label
+            else:
+                label = option.metric_label
+            self._api_value_to_label_map[option.api_value] = label
+            self._label_to_api_value_map[label] = option.api_value
+
+        self._attr_options = list(self._label_to_api_value_map)
+
+    async def async_select_option(self, option: str) -> None:
+        """Change the selected option."""
+        try:
+            await self._data.controller.restrictions.set_universal(
+                {self.entity_description.data_key: self._label_to_api_value_map[option]}
+            )
+        except RainMachineError as err:
+            LOGGER.error("Error while setting %s: %s", self.name, err)
+
+    @callback
+    def update_from_latest_data(self) -> None:
+        """Update the entity when new data is received."""
+        raw_value = self.coordinator.data[self.entity_description.data_key]
+
+        try:
+            label = self._api_value_to_label_map[raw_value]
+        except KeyError:
+            LOGGER.error("Cannot translate raw value: %s", raw_value)
+            label = None
```

This looks like it would be a bug for KeyError to happen?  Should we let this blow up?

---

Good point: I probably don't need to be so weirdly defensive here.